In [3]:
import pandas as pd

In [4]:
import matplotlib as plt

In [5]:
import pandas as pd

In [ ]:
data=pd.read_table("C:\\Users\\Muskan Sinha\\Downloads\\P176.txt")

In [7]:
data.head()

,X,Y
0,294,30
1,247,32
2,267,37
3,358,44
4,423,47


In [8]:
x= data['X']
y= data['Y']
y


0      30
1      32
2      37
3      44
4      47
5      49
6      56
7      62
8      68
9      78
10     80
11     84
12     88
13     97
14    100
15    109
16    114
17    117
18    106
19    128
20    130
21    160
22     97
23    180
24    112
25    210
26    135
Name: Y, dtype: int64

In [9]:
import numpy as np


In [10]:
np.sum(x)

np.int64(20500)

In [11]:
np.sum(x*y)

np.int64(2323945)

In [12]:
np.sum(x*x)

np.int64(19245812)

In [13]:
data.shape

(27, 2)

In [14]:
n=data.shape[0]

In [15]:
beta1=(n*np.sum(x*y)-np.sum(x)*np.sum(y))/(n*np.sum(x*x)-(np.sum(x)**2))
beta1

np.float64(0.10536109357806465)

In [16]:
beta0=1/n*(np.sum(y)-beta1*(np.sum(x)))
beta0

np.float64(14.448058579617584)

Linear Regression is
Y=14.448 + 0.105*X

## Fit a linear regression model for multiple variable

In [ ]:
data_X= pd.read_csv("C:\\Users\\Muskan Sinha\\Downloads\\LR_X.csv")

In [18]:
data_X.head()

,X1,X2,X3
0,0.26,0.08,0.28
1,0.26,0.52,0.01
2,0.85,0.60,0.14
3,0.94,0.22,0.48
4,0.53,0.05,0.76


In [19]:
data_X.shape

(1000, 3)

In [ ]:
data_Y= pd.read_csv("C:\\Users\\Muskan Sinha\\Downloads\\LR_Y.csv")

In [21]:
data_Y.head()

,Y
0,1.472
1,1.802
2,1.548
3,1.390
4,2.141


In [22]:
data_Y.head

<bound method NDFrame.head of          Y
0    1.472
1    1.802
2    1.548
3    1.390
4    2.141
..     ...
995  2.245
996  3.471
997  0.796
998  2.406
999  1.870

[1000 rows x 1 columns]>

In [23]:
data_X.transpose()

,0,1,2,3,4,5,6,7,8,9,...,990,991,992,993,994,995,996,997,998,999
X1,0.26,0.26,0.85,0.94,0.53,0.07,0.99,0.22,0.61,0.29,...,0.84,0.69,0.22,0.90,0.48,0.78,0.76,0.88,0.28,0.94
X2,0.08,0.52,0.60,0.22,0.05,0.59,0.94,0.39,0.52,0.31,...,0.74,0.99,0.59,0.24,0.40,0.41,0.71,0.06,0.40,0.48
X3,0.28,0.01,0.14,0.48,0.76,0.63,0.96,0.52,0.36,0.01,...,0.96,0.81,0.03,0.46,0.93,0.60,0.86,0.33,0.41,0.45


In [27]:
beta2=(data_X.transpose()*data_X)**-1*(data_X.transpose()*data_Y)

beta2

,0,1,2,3,4,5,6,7,8,9,...,994,995,996,997,998,999,X1,X2,X3,Y
X1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
X2,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
X3,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
995,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
996,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
997,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
998,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [25]:
data_X.shape

(1000, 3)

In [26]:
data_Y.shape

(1000, 1)

In [28]:
ones=np.ones((1000,1))

In [29]:
ones.shape

(1000, 1)

In [30]:
ones[:10]

array([[1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.]])

In [32]:
new_X=np.concatenate((ones,data_X),axis=1)
new_X

array([[1.  , 0.26, 0.08, 0.28],
       [1.  , 0.26, 0.52, 0.01],
       [1.  , 0.85, 0.6 , 0.14],
       ...,
       [1.  , 0.88, 0.06, 0.33],
       [1.  , 0.28, 0.4 , 0.41],
       [1.  , 0.94, 0.48, 0.45]])

In [33]:
res1=np.linalg.inv(np.dot(np.transpose(new_X),new_X))

In [34]:
res1.shape

(4, 4)

In [36]:
res2=np.dot(np.transpose(new_X),data_Y)
res2.shape

(4, 1)

In [37]:
beta=np.dot(res1,res2)
print(beta)

[[ 1. ]
 [-1.2]
 [ 2.1]
 [ 2.2]]


# The equation of the linear regression is Y=1.0-1.2x1+2.1x2+2.2*x2


In [41]:
data_X

,X1,X2,X3
0,0.26,0.08,0.28
1,0.26,0.52,0.01
2,0.85,0.60,0.14
3,0.94,0.22,0.48
4,0.53,0.05,0.76
...,...,...,...
995,0.78,0.41,0.60
996,0.76,0.71,0.86
997,0.88,0.06,0.33
998,0.28,0.40,0.41


# Gradient Descent

In [43]:
data

,X,Y
0,294,30
1,247,32
2,267,37
3,358,44
4,423,47
5,311,49
6,450,56
7,534,62
8,438,68
9,697,78


In [42]:
x=data['X']
y=data['Y']

# Normalize your data

In [45]:
x1=(x-np.mean(x))/np.std(x)
y1=(y-np.mean(y))/np.std(y)

In [46]:
def grad_beta0(beta0,beta1,x1,y1):
    n=len(x1)
    res=0
    for i in range(n):
        res+=(y1[i]-beta0-beta1*x1[i])
    grad_b0=-2/float(n)*res
    #print(grad_b0)
    return grad_b0